### ISSUE TITLE
Go Advent Day 24 - Channel Buffering Patterns GopherAcademy

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4470]
Go Advent Day 24 - Channel Buffering Patterns GopherAcademy<br>
<br>
http://bit.ly/2RqmHOd<br>
<br>
About 0 Minutes<br>
<br>
<div>
<h2>Go Advent Day 24 - Channel Buffering Patterns</h2>
<br>
<h2>Introduction</h2>
<p>One common method for message processing in Go is to receive from an input channel and write to an output channel, often using intermediate channels for message transformation and filtering. Multiple goroutines may perform these functions concurrently and independently, making for code that is easily parallelized and tested.</p>
<p>Message buffering is one kind of transformation that is sometimes useful in these systems. Some programs don&rsquo;t need to process each message immediately, and can more efficiently process several messages at once. Other programs receive bursty input but are able to coalesce groups of messages.</p>
<p>An example of the latter case is software built with an inotify/kqueue library such as <a href="https://github.com/howeyc/fsnotify">github.com/howeyc/fsnotify</a>. This library provides the user with a channel of file events. Messages from fsnotify often come in bursts; for instance, some text editors save files by writing out a temporary file and moving it into place &ndash; a create event, one or more modifications, and a rename.</p>
<p>I observed such behavior when writing <a href="https://github.com/cespare/reflex">github.com/cespare/reflex</a>, a tool for running commands when files change, and the code below is similar to <a href="https://github.com/cespare/reflex/blob/master/workers.go">real code in that program</a>.</p>
<p>In this post I will explore how to handle bursty messages in channel pipelines.</p>
<h2>Producing messages</h2>
<p>We&rsquo;ll represent messages with <code>Event</code> s, which are just <code>int</code> s. Multiple <code>Event</code> s are allowed to be merged into a single event using addition. (In an fsnotify-based program, for example, you might be able to coalesce multiple events by preserving only the unique file names, or only the fact that some file was changed at all.) Our producer intermittently emits a burst of random-valued <code>Event</code> s.</p>
<pre>
<code>type Event int

func NewEvent() Event                   { return 0 }
func (e Event) Merge(other Event) Event { return e + other }

func produce(out chan&lt;- Event) {
    for {
        delay := time.Duration(rand.Intn(5)+1) * time.Second
        nMessages := rand.Intn(10) + 1
        time.Sleep(delay)
        for i := 0; i &lt; nMessages; i++ {
            out &lt;- Event(rand.Intn(10))
        }
    }
}
</code>
</pre>
<p>If we print out all the messages created by <code>produce</code>, we can see that they come out in small batches. <a href="http://play.golang.org/p/aHsHNFIfGP">See the runnable playground example</a> (you may have to scroll down to see the output).</p>
<h2>Batching by time</h2>
<p>Let&rsquo;s suppose we are willing to wait for a short time after receiving a message to see if other messages arrive that we can coalesce with the first. One simple way of doing this is simply to send every N milliseconds and merge all received messages in the meantime. This is easy with a <a href="http://golang.org/pkg/time/#Ticker"><code>time.Ticker</code></a>:</p>
<pre>
<code>func coalesce500(in &lt;-chan Event, out chan&lt;- Event) {
    ticker := time.NewTicker(500 * time.Millisecond)
    event := NewEvent()
    for {
        select {
        case &lt;-ticker.C:
            out &lt;- event
            event = NewEvent()
        case e := &lt;-in:
            event = event.Merge(e)
        }
    }
}
</code>
</pre>
<p><a href="http://play.golang.org/p/yelQeZiyly">See the full runnable code here</a>. This works, but it has a few problems:</p>
<ul>
<li>If no messages are arriving, every 500ms <code>coalesce500</code> sends an empty message.</li>
<li>If a message arrives shortly before the ticker fires, we&rsquo;ll send it by itself without waiting for more messages. (More generally, a burst falling across the ticker will create two messages, rather than one.)</li>
<li>If the receiver doesn&rsquo;t handle the message immediately, the coalescing goroutine will be blocked, which in turns blocks the upstream producer. (In real code, that producer may be constructed to discard any messages it cannot immediately send.)</li>
</ul>
<p>Let&rsquo;s tackle the third problem first.</p>
<h2>Handling slow receivers</h2>
<p>One obvious way to avoid blocking on a slow receiver is to change the output channel to be buffered. This only helps until the buffer fills up, though, and furthermore we won&rsquo;t merge multiple buffered messages, potentially causing a downstream burst as well. A better way of handling the slow receiver is to implement the desired semantics directly: keep coalescing messages until the receiver is ready to accept more input.</p>
<pre>
<code>func coalesceSlow(in &lt;-chan Event, out chan&lt;- Event) {
    event := NewEvent()
    for e := range in {
        event = event.Merge(e)
    loop:
        for {
            select {
            case e := &lt;-in:
                event = event.Merge(e)
            case out &lt;- event:
                event = NewEvent()
                break loop
            }
        }
    }
}
</code>
</pre>
<p>You can <a href="http://play.golang.org/p/bYZpWB974A">run the code with a slow receiver</a> and see that this has the desired effect: we only create a single coalesced message while the receiver blocks. <code>coalesceSlow</code> still suffers from the problem as <code>coalesce500</code> that it does not wait for a follow-up burst after receiving a message. Further, it introduces a bit of ugly code: we receive from <code>in</code> in two different places (the range and inside the select) and we have to bounce between the inner and outer loops depending on whether we have anything to send.</p>
<h2>Putting it together</h2>
<p>To review, we would like for our final version to do the following:</p>
<ul>
<li>Wait for 500ms after the receipt of a message to send</li>
<li>Wait until the receiver is ready before sending</li>
<li>Keep coalescing incoming messages while waiting</li>
<li>Only have single channel receive and send statements instead of duplicates</li>
</ul>
<p>Our state machine will need to be a little more complex to meet these goals. To solve the empty message problem from <code>coalesce500</code>, we&rsquo;ll start a timer running only after receiving an event. Then we&rsquo;ll coalesce messages until the timer fires. Finally, we&rsquo;ll continue coalescing messages until the output channel is ready, after which we can send and go back to the first state again.</p>
<p>Here is a diagram to show these transitions:</p>
<p><img src="https://blog.gopheracademy.com/postimages/day-24-channel-buffering-patterns/channel-buffering.png"></p>
<p>Here is the code:</p>
<pre>
<code>func coalesce(in &lt;-chan Event, out chan&lt;- Event) {
    event := NewEvent()
    timer := time.NewTimer(0)

    var timerCh &lt;-chan time.Time
    var outCh chan&lt;- Event

    for {
        select {
        case e := &lt;-in:
            event = event.Merge(e)
            if timerCh == nil {
                timer.Reset(500 * time.Millisecond)
                timerCh = timer.C
            }
        case &lt;-timerCh:
            outCh = out
            timerCh = nil
        case outCh &lt;- event:
            event = NewEvent()
            outCh = nil
        }
    }
}
</code>
</pre>
<p>There are a few techniques used here that deserve explanation. Most notably, we&rsquo;re exploiting the fact that any cases inside a select which receive from or send to a nil channel are skipped (read more about this in the <a href="http://golang.org/ref/spec#Select_statements">language spec</a>). We use this in the second and third cases: <code>timerCh</code> and <code>outCh</code> are toggled between nil and the <code>timer.C</code> / <code>out</code> channels to selectively remove them from the select.</p>
<p>In the initial state (labeled &ldquo;ready&rdquo; in the graph), both <code>timerCh</code> and <code>outCh</code> are nil, so we simply block on <code>in</code>, waiting for input. After we receive a message, the timer is set for 500ms in the future and <code>timerCh</code> is set to the timer&rsquo;s output channel; further messages are coalesced while we wait for the timer to fire (the second state, &ldquo;waiting&rdquo;). Once this happens, we toggle <code>outCh</code> to <code>out</code>, so that we can detect when the output channel is ready to receive (this is the last state, &ldquo;ready to send&rdquo;). Once we have successfully sent, we return to the initial state, <code>outCh</code> and <code>timerCh</code> having previously been reset to nil.</p>
<p>The <code>timer</code> variable initialized with <code>time.NewTimer(0)</code> fires immediately, but we don&rsquo;t start listening to its output channel until after we have <code>Reset</code> it. (There is no way to create a <code>time.Timer</code> without scheduling it.)</p>
<p>If you <a href="http://play.golang.org/p/r4nH9M38Jt">run this on the playground</a> you can verify that it behaves as advertised.</p>
<h2>The end</h2>
<p>If you have questions, you can find me on <a href="https://plus.google.com/+CalebSpare/">Google+</a> or <a href="https://twitter.com/kingfishr">Twitter</a>. I am also cespare on IRC; if you haven&rsquo;t checked it out, we have a great community at #go-nuts on Freenode.</p>
<p>Special thanks to Tommi Virtanen and Dominik Honnef for the excellent diagram and valuable feedback as I was editing this post.</p>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via blog.gopheracademy.com http://bit.ly/1U7abwd<br>
<br>
June 21, 2019 at 04:14PM
